migrations/back_37/back_37.go

```diff
@@ -0,0 +1,244 @@
+	defer cancel()
+	application.RunAndExit(NewMigration(ctx))
+}
+
+type Migration struct {
+	ctx context.Context
+	*migrationMongo.Migration
+	dataRepository *storeStructuredMongo.Repository
+}
+
+func getValidatedString(bsonData bson.M, fieldName string) (string, error) {
+	if valRaw, ok := bsonData[fieldName]; !ok {
+		return "", errors.Newf("%s is missing", fieldName)
+	} else if val, ok := valRaw.(string); !ok {
+		return "", errors.Newf("%s is not of expected type", fieldName)
+	} else if val == "" {
+		return "", errors.Newf("%s is empty", fieldName)
+	} else {
+		return val, nil
+	}
+}
+
+func getValidatedTime(bsonData bson.M, fieldName string) (time.Time, error) {
+	if valRaw, ok := bsonData[fieldName]; !ok {
+		return time.Time{}, errors.Newf("%s is missing", fieldName)
+	} else if val, ok := valRaw.(time.Time); !ok {
+		return time.Time{}, errors.Newf("%s is not of expected type", fieldName)
+	} else if val.IsZero() {
+		return time.Time{}, errors.Newf("%s is empty", fieldName)
+	} else {
+		return val, nil
+	}
+}
+
+func createDatumHash(bsonData bson.M) (string, error) {
+	identityFields := []string{}
+	if userID, err := getValidatedString(bsonData, "_userId"); err != nil {
+		return "", err
+	} else {
+		identityFields = append(identityFields, userID)
+	}
+	if deviceID, err := getValidatedString(bsonData, "deviceId"); err != nil {
+		return "", err
+	} else {
+		identityFields = append(identityFields, deviceID)
+	}
+	if datumTime, err := getValidatedTime(bsonData, "time"); err != nil {
+		return "", err
+	} else {
+		identityFields = append(identityFields, datumTime.Format(types.TimeFormat))
+	}
+	datumType, err := getValidatedString(bsonData, "type")
+	if err != nil {
+		return "", err
+	}
+	identityFields = append(identityFields, datumType)
+
+	switch datumType {
+	case "basal":
+		if deliveryType, err := getValidatedString(bsonData, "deliveryType"); err != nil {
+			return "", err
+		} else {
+			identityFields = append(identityFields, deliveryType)
+		}
+	case "bolus", "deviceEvent":
+		if subType, err := getValidatedString(bsonData, "subType"); err != nil {
+			return "", err
+		} else {
+			identityFields = append(identityFields, subType)
+		}
+	case "smbg", "bloodKetone", "cbg":
+		if units, err := getValidatedString(bsonData, "units"); err != nil {
+			return "", err
+		} else {
+			identityFields = append(identityFields, units)
+		}
+
+		valueRaw, ok := bsonData["value"]
+		if !ok {
+			return "", errors.New("value is missing")
+		}
+		val, ok := valueRaw.(float64)
+		if !ok {
+			return "", errors.New("value is not of expected type")
+		}
+
+		if len(fmt.Sprintf("%v", valueRaw)) > 7 {
+			// NOTE: we need to ensure the same precision for the
+			// converted value as it is used to calculate the hash
+			mgdlVal := val*18.01559 + 0.5
+			mgdL := glucose.MgdL
+			val = *glucose.NormalizeValueForUnits(&mgdlVal, &mgdL)
+		}
+		strVal := strconv.FormatFloat(val, 'f', -1, 64)
+		identityFields = append(identityFields, strVal)
```

Comment:
ok - I have reworked that.

Reply:
> I think what this means is that we'll need to ensure the new Uploader code that replaces jellyfish with platform will either 1) send mg/dL values, or 2) send mmol/L values and truncate to 6 digits in the Uploader before uploading. So, if the Uploader will upload mmol/L values NOT truncated to 6 digits for some devices, then this code here will need to change. My concern here is that we'd end up with a different deduplicate hash from the one after the Uploader is sending to platform.

So wouldn't we just do the same as other devices the uploader ingests data from?  From my understanding 
- it takes the units and value as whatever the devices gives us and then the platform normalize's that to save as `mmol/L` units and converts the value (if it wasn't mmol/L).
- the platform truncation would occur that same as any other device that is uploaded
- the hash is all good as both our JF hash and migration hash use a truncated value

Reply:
Yes, all of the above is correct.

I think the problem is in the second bullet, "the platform truncation would occur that same as any other device that is uploaded". Specifically, from `glucose.go` in `platform`:
```
func NormalizeValueForUnits(value *float64, units *string) *float64 {
	if value != nil && units != nil {
		switch *units {
		case MgdL, Mgdl:
			intValue := int(*value/MmolLToMgdLConversionFactor*MmolLToMgdLPrecisionFactor + 0.5)
			floatValue := float64(intValue) / MmolLToMgdLPrecisionFactor
			return &floatValue
		}
	}
	return value
}
```

Note that the truncation **only** occurs when converting from mg/dL to mmol/L and does **not** occur if the units uploaded is already mmol/L. Thus, for this migration to be consistent with any new uploader code that moves from Jellyfish to Platform uploads, the new Uploader code has to conform to either 1) send mg/dL values (and above normalization code will truncate), or 2) send mmol/L values already truncated to 6 digits using the same formula.

For example, with this migration code, **all** mmol/L values will be truncated. This will be the same **iff** the new Uploader code sends mg/dL or does the truncation itself. However, if the new Uploader code sends non-truncated mmol/L values, the `platform` will not do any truncation (see `NormalizeValueForUnits` above) and therefore the deduplicator hash generated by this migration code (which has truncated all values) will not match the deduplicator hash generated when the new Uploader code uploads non-truncated mmol/L values.

Does that make sense?